### ISSUE TITLE
SkiaSharp does not work in Docker Windows Container.

### ISSUE THREAD

**Issue by AlexNosk** [mono/SkiaSharp#591]
### Description

SkiaSharp throws when run in Windows container.

### Code

Steps to reproduce:
1.	Create a simple Console application that uses SkiaSharp

>using SkiaSharp;
>using System;
>
>namespace SkiaSharpDockerTest
>{
>    class Program
>    {
>        static void Main(string[] args)
>        {
>            Console.WriteLine(Environment.OSVersion);
>
>            try
>            {
>                using (SKBitmap bmp = new SKBitmap(100, 100))
>                using (SKCanvas canvas = new SKCanvas(bmp))
>                using (SKPaint p = new SKPaint())
>                {
>                    p.Color = SKColors.Red;
>                    canvas.DrawCircle(50, 50, 50, p);
>
>                    using (SKFileWStream fs = new SKFileWStream("out.png"))
>                        bmp.Encode(fs, SKEncodedImageFormat.Png, 100);
>                }
>            }
>            catch (Exception ex)
>            {
>                Console.WriteLine(ex.Message);
>                Console.WriteLine(ex.StackTrace);
>            }
>    }
>}

2.	Create a Dokerfile

>FROM microsoft/dotnet:2.0-sdk
>WORKDIR /app
>COPY . ./
>RUN dotnet publish -c Release -o out
>ENTRYPOINT ["dotnet", "out/SkiaSharpDockerTest.dll"]

3.	Build and run image
> docker build -t skiasharptest .
> docker run --rm skiasharptest SkiaSharpDockerTest from Docker

As a result exception is thrown
The type initializer for 'SkiaSharp.SKImageInfo' threw an exception.
   at SkiaSharp.SKBitmap..ctor(Int32 width, Int32 height, Boolean isOpaque)
   at AsposeWordsDockerTest.Program.Main(String[] args) in C:\app\Program.cs:line 15

### Expected Behavior

Red circle image should be created

### Actual Behavior

throws an exception

### Basic Information

- Version with issue:  1.60.2
- Last known good version:  none
- IDE:  Visual Studio
- Platform Target Frameworks: .NET Core 2.0
- Target Devices: Docker Windows container

**Comment by AlexNosk** [comment]
Any update on this?

**Comment by kekekeks** [comment]
@AlexNosk 

You are probably missing Visual C++ runtime. See https://github.com/mono/SkiaSharp/issues/136

**Comment by chrismellard** [comment]
@kekekeks This is one issue with running SkiaSharp inside Windows containers but not the only one. libSkiaSharp.dll also has a dependency on both opengl32.dll and glu32.dll. 

@AlexNosk In your dockerfile copy these two assemblies in to the windows/system32 folder and things work fine - least they did for me. This is a hack, and so long as you don't any of whatever skiasharp requires from those libraries to actually work you're all good. I think they're working on 3d support for windows containers at the moment.

**Comment by kekekeks** [comment]
On Linux opengl references are resolved dynamically. See  https://github.com/mono/skia/pull/50 and
https://github.com/mono/skia/commit/b5793908b1fa1510c31cf7548035d7df0566f7bc

I guess on Windows we can do the same.

**Comment by mattleibow** [comment]
@kekekeks I am thinking that maybe we should - it appears that those assemblies are not always distributed on some "versions"... What are they actually? @chrismellard what container were you using? I thought they were always part of all Windows.

**Comment by chrismellard** [comment]
@mattleibow Container I had to copy those libraries in from was 

`FROM microsoft/iis:windowsservercore-1803`

**Comment by mattleibow** [comment]
After that commit, the new dependency tree is basically:
 - fontsub.dll
 - ole32.dll
 - user32.dll
 - kernel32.dll
 - msvcp140.dll _(vcredist)_
 - vcruntime140.dll _(vcredist)_

![deps](https://user-images.githubusercontent.com/1096616/46575612-c02adc80-c9b8-11e8-9b86-9c834009cb2d.png)

I am also working on making sure that there is an easy path for Docker. I have managed top basically get this working:

```
RUN Invoke-WebRequest -OutFile vc_redist.x64.exe https://aka.ms/vs/15/release/vc_redist.x64.exe `
    Start-Process vc_redist.x64.exe -ArgumentList '/install /passive /norestart' -Wait
```

I will be creating a wiki with all the steps for creating a Docker container for both Windows and Linux. 

Right now, nanoserver is not playing nice and several dependencies are missing, so I don't think that support will be coming soon, but it will work fine with windowsservercore variants.

**Comment by kekekeks** [comment]
May be it's time to reconsider about static CRT linking? It doesn't increase binary size that much, but decreases surprise factor by a lot.

**Comment by mattleibow** [comment]
Having a look at Nano Server... Ran the `NanoServerApiScan.exe` tool from: https://blogs.technet.microsoft.com/nanoserver/2016/04/27/nanoserverapiscan-exe-updated-for-tp5/

```
=== libSkiaSharp.dll ===

ERRORS:

  FONTSUB.dll
    CreateFontPackage(Proc not found)
  USER32.dll
    SystemParametersInfoW(Proc not found)
  MSVCP140.dll
    ?xsputn@?$basic_streambuf@DU?$char_traits@D@std@@@std@@MEAA_JPEBD_J@Z(Proc not found)
    ?xsgetn@?$basic_streambuf@DU?$char_traits@D@std@@@std@@MEAA_JPEAD_J@Z(Proc not found)
    ?uflow@?$basic_streambuf@DU?$char_traits@D@std@@@std@@MEAAHXZ(Proc not found)
    ?sync@?$basic_streambuf@DU?$char_traits@D@std@@@std@@MEAAHXZ(Proc not found)
    ?showmanyc@?$basic_streambuf@DU?$char_traits@D@std@@@std@@MEAA_JXZ(Proc not found)
    ?setbuf@?$basic_streambuf@DU?$char_traits@D@std@@@std@@MEAAPEAV12@PEAD_J@Z(Proc not found)
    ?imbue@?$basic_streambuf@DU?$char_traits@D@std@@@std@@MEAAXAEBVlocale@2@@Z(Proc not found)
    ?_Unlock@?$basic_streambuf@DU?$char_traits@D@std@@@std@@UEAAXXZ(Proc not found)
    ?_Lock@?$basic_streambuf@DU?$char_traits@D@std@@@std@@UEAAXXZ(Proc not found)
    ??1?$basic_iostream@DU?$char_traits@D@std@@@std@@UEAA@XZ(Proc not found)
    ??0?$basic_iostream@DU?$char_traits@D@std@@@std@@QEAA@PEAV?$basic_streambuf@DU?$char_traits@D@std@@@1@@Z(Proc not found)
    ??5?$basic_istream@DU?$char_traits@D@std@@@std@@QEAAAEAV01@AEAN@Z(Proc not found)
    ??6?$basic_ostream@DU?$char_traits@D@std@@@std@@QEAAAEAV01@_K@Z(Proc not found)
    ??6?$basic_ostream@DU?$char_traits@D@std@@@std@@QEAAAEAV01@_J@Z(Proc not found)
    ??0?$basic_ios@DU?$char_traits@D@std@@@std@@IEAA@XZ(Proc not found)
    ?imbue@?$basic_ios@DU?$char_traits@D@std@@@std@@QEAA?AVlocale@2@AEBV32@@Z(Proc not found)
    ??1?$basic_ios@DU?$char_traits@D@std@@@std@@UEAA@XZ(Proc not found)
    ?_Pninc@?$basic_streambuf@DU?$char_traits@D@std@@@std@@IEAAPEADXZ(Proc not found)
    ??1?$basic_streambuf@DU?$char_traits@D@std@@@std@@UEAA@XZ(Proc not found)
    ??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@IEAA@XZ(Proc not found)
    ?classic@locale@std@@SAAEBV12@XZ(Proc not found)
    ?_Xbad_alloc@std@@YAXXZ(Proc not found)
    ?_Xlength_error@std@@YAXPEBD@Z(Proc not found)
    _Query_perf_frequency(Proc not found)
    _Query_perf_counter(Proc not found)
    ?_Xbad_function_call@std@@YAXXZ(Proc not found)
      Please copy Nano Server compliant MSVCP140.dll from %ProgramFiles(x86)%\Microsoft Visual Studio 14.0\VC\redist\onecore\x64\Microsoft.VC140.CRT\
  VCRUNTIME140.dll
    __std_type_info_destroy_list(Proc not found)
    __std_exception_destroy(Proc not found)
    __std_exception_copy(Proc not found)
    _purecall(Proc not found)
    __std_terminate(Proc not found)
    __CxxFrameHandler3(Proc not found)
    __std_type_info_compare(Proc not found)
    memcpy(Proc not found)
    memmove(Proc not found)
    memcmp(Proc not found)
    strchr(Proc not found)
    longjmp(Proc not found)
    strstr(Proc not found)
    _CxxThrowException(Proc not found)
    strrchr(Proc not found)
    __C_specific_handler(Proc not found)
    __intrinsic_setjmp(Proc not found)
      Please copy Nano Server compliant VCRUNTIME140.dll from %ProgramFiles(x86)%\Microsoft Visual Studio 14.0\VC\redist\onecore\x64\Microsoft.VC140.CRT\
```

**Comment by mattleibow** [comment]
Just did some investigation :)

`CreateFontPackage` from FONTSUB.dll is just for XPS support - and this is also just for font subsetting. The UWP build does not have this at all, so I could just remove it for a nano build.

`SystemParametersInfoW` from USER32.dll is also not too essential - it is just for determining the default font. Again, this is not available for UWP - we just pick the "Segoe UI" which we know is available by default.

Nano Server support could very well be possible in the near future - just have to check priorities.

**Comment by mattleibow** [comment]
@kekekeks I have been thinking about this and I will link in #136 so we can keep the discussion linked, but that discussion should be done there. (going to make a comment there now)

**Comment by mattleibow** [comment]
Closing this as I have done a few things:
 - the Windows Nano Server image is not yet supported. See my comment https://github.com/mono/SkiaSharp/issues/591#issuecomment-427605252 and https://github.com/mono/SkiaSharp/issues/591#issuecomment-427605722
 - the new version of SkiaSharp will be statically linked #136 (@kekekeks)

I have opened a new issue to support Nano server: #676

**Comment by mattleibow** [comment]
Nano server images are now supported after #1040

**Comment by mattleibow** [comment]
Just linking this new issue : https://github.com/NuGet/Home/issues/8894

**Comment by djarvis** [comment]
I ended up here because of some confusion over SkiaSharp in a NanoServer docker container.  While I was able to get some things working after including SkiaSharp.NativeAssets.NanoServer, I noticed that I cannot draw text onto a bitmap using a font that I load from a file:

`
var typeFace = SKTypeface.FromFile("Font Awesome 5 Free-Solid-900.otf");
var paint = new SKPaint() { TextSize = 40, Color = SKColor.Parse("FF0000"), Typeface = typeFace};
SKRect bounds = new SKRect();
paint.MeasureText(unicodeCharacter, ref bounds);
`

bounds is always 0x0.

I see that SkiaSharp.NativeAssets.NanoServer shows:

_This variation of the Windows native assets includes the build of libSkiaSharp.dll that does not make use of the typeface subsetting when creating XPS documents (CreateFontPackage from FONTSUB.dll)._

So I'm guessing this has to do with something.

Is there any way to use a font in NanoServer?

**Comment by Gillibald** [comment]
Unfortunately there is currently no support for text rendering on Nano Server. That will most likely change with the next major release of SkiaSharp when Freetype can be used for Windows builds.

**Comment by mattleibow** [comment]
@djarvis as @Gillibald says, there is currently an issue #1105

For now, is it possible to use the Linux images? Typically they are smaller and just as good. And, if you have a Windows host, they can be run side-by-side with Windows containers. That is the best for now, or use a Windows Server Core image - [like this example](https://github.com/mono/SkiaSharp/tree/master/samples/Basic/Docker/WebApi).

**Comment by djarvis** [comment]
@mattleibow We have been using solely Linux containers but now have to offer up Windows containers.  I'd rather not have the 4.8GB footprint just to be able to draw a font onto a bitmap so I'd like to stick to the nanoserver images.

We need to support simple, single-machine deployments.  With the need to run along side legacy Windows apps, we see Windows Server 2019 as a solution since it comes with a free Docker EE license and as of now LCOW is still in an unsupported preview.  Otherwise we'd just continue with Linux containers.

I just need to draw this one font onto a bitmap so I might be able to take the svg and find a renderer and draw it manually but I haven't looked too far into that.

**Comment by Gillibald** [comment]
You will be soon able to generate glyph outlines with the help of HarfBuzzSharp. Maybe thats a good way to get text rendered without having to rely on a font host implementation.

**Comment by djarvis** [comment]
Luckily since I just needed to draw the Font Awesome glyphs I was able to use SkiaSharp.Svg to draw the relatively simple svgs onto a bitmap
  This seems to work on Linux and Windows.

**Comment by christian289** [comment]
Hello.
I am trying to use SkiaSharp in windows docker container.
My environment is below.
- .NET 5.
- Windows 10 Pro, 21H1, 19043.1348
- Visual Studio 2019 Pro
- SkiaSharp.NativeAsserts.Linux.NoDependencies 2.80.3

Below is a screenshot of the Docker desktop.
![image](https://user-images.githubusercontent.com/8288586/141957804-4f2c31c9-bb35-4b14-9306-7705cf8f4f18.png)

--------------------------------------------------------------------------------------------------------------------------------------------
warn]0;C:\Program Files\dotnet\dotnet.exewarn: ImageAutoSliceBackgroundServiceForWindows.ImageAutoSliceWorker[0] 
      [11/16/2021 6:17:10 PM][ImageAutoSliceWorker]Input Message        
System.TypeInitializationException: The type initializer for 'SkiaSharp.SKAbstractManagedStream' threw an exception.
 ---> System.DllNotFoundException: Unable to load DLL 'libSkiaSharp' or one of its dependencies: The specified module could not be found. (0x8007007E)
   at SkiaSharp.SkiaApi.sk_managedstream_set_procs(SKManagedStreamDelegates procs)
   at SkiaSharp.SKAbstractManagedStream..cctor()
   --- End of inner exception stack trace ---
   at SkiaSharp.SKAbstractManagedStream..ctor(Boolean owns)
   at SkiaSharp.SKManagedStream..ctor(Stream managedStream, Boolean disposeManagedStream)
   at ImageAutoSliceBackgroundServiceForWindows.ImageAutoSliceWorker.MakeSKBitmap(ReadOnlySpan`1 span) in C:\app\Workers\ImageAutoSliceWorker.cs:line 192       
   at ImageAutoSliceBackgroundServiceForWindows.ImageAutoSliceWorker.MessageHandler(ProcessMessageEventArgs args)
   at Azure.Messaging.ServiceBus.ServiceBusProcessor.OnProcessMessageAsync(ProcessMessageEventArgs args)
   at Azure.Messaging.ServiceBus.ReceiverManager.OnMessageHandler(ServiceBusReceivedMessage message, CancellationToken processorCancellationToken)
   at Azure.Messaging.ServiceBus.ReceiverManager.ProcessOneMessage(ServiceBusReceivedMessage message, CancellationToken cancellationToken)
--------------------------------------------------------------------------------------------------------------------------------------------

The above problem occurs when the first class related to SkiaSharp is called. The code to be called is as follows.

```
        private SKBitmap MakeSKBitmap(ReadOnlySpan<byte> span) // 189 line
        {                                                                                                   // 190 line
            MemoryStream ms = new(span.ToArray());                             // 191 line
            SKManagedStream skStream = new(ms, false);                      // 192 line, Exception Occured !!
            SKCodec codec = SKCodec.Create(skStream);

            return SKBitmap.Decode(codec);
        }
```

Dockerfile
```
FROM mcr.microsoft.com/dotnet/sdk:5.0 AS build
SHELL ["pwsh", "-Command", "$ErrorActionPreference = 'Stop'; $ProgressPreference = 'SilentlyContinue';"] #powershell은 Desktop Edition이고, pwsh은 Core Edition이다., https://docs.microsoft.com/ko-kr/powershell/module/microsoft.powershell.core/about/about_preference_variables?view=powershell-7.2
USER ContainerAdministrator
RUN Invoke-WebRequest https://raw.githubusercontent.com/microsoft/artifacts-credprovider/master/helpers/installcredprovider.ps1 -OutFile ./installcredprovider.ps1; 
RUN .\installcredprovider.ps1;
RUN del installcredprovider.ps1
USER ContainerUser
WORKDIR /app
COPY ["ImageAutoSliceBackgroundServiceForWindows.csproj", ""]
COPY ["nuget.config", ""]

ARG ID
ARG PAT
RUN dotnet restore

COPY . .
RUN dotnet build -c Release -o out --no-restore # 이 부분에서 .dockerignore가 없으면 MSB4018 에러가 발생한다. https://github.com/dotnet/dotnet-docker/issues/1649#issuecomment-581433906
RUN dotnet publish

FROM mcr.microsoft.com/dotnet/runtime:5.0 AS runtime
WORKDIR /app/
COPY --from=build /app/out .

COPY /app/runtimes/win-x64/native
ENTRYPOINT ["dotnet", "ImageAutoSliceBackgroundServiceForWindows.dll"]
```

nuget.config
```
<?xml version="1.0" encoding="utf-8"?>
<configuration>
  <packageSources>
    <clear />
    <add key="nuget.org" value="https://api.nuget.org/v3/index.json" />
    <add key="EcremmoceSDK" value="https://pkgs.dev.azure.com/ecremmoce/_packaging/EcremmoceSDK/nuget/v3/index.json" />
  </packageSources>
  <activePackageSource>
    <add key="All" value="(Aggregate source)" />
  </activePackageSource>
  <packageSourceCredentials>
    <EcremmoceSDK>
      <add key="Username" value="%ID%" />
      <add key="ClearTextPassword" value="%PAT%" />
    </EcremmoceSDK>
  </packageSourceCredentials>
</configuration>
```
If you look at dockerfile and nuget.config, there are parts that are slightly different from the basic contents. This is all part of linking Azure DevOps Artifact and PAT, so it doesn't seem like it has anything to do with SkiaSharp, but I share it nonetheless.

I used SkiaSharp when developing WPF in .NET 5. At that time, I developed using 'SkiaSharp.NativeAsserts.Linux.NoDependencies 2.80.3', and even MSIX packaging was successful. So, again, I tried the following to use it as a Windows Container.

But the results were not good.
What more can I do?
